component/file_cache/file_cache.go

```diff
@@ -395,30 +399,48 @@ func (fc *FileCache) invalidateDirectory(name string) {
 		return
 	}
 	// TODO : wouldn't this cause a race condition? a thread might get the lock before we purge - and the file would be non-existent
+	var directoryStack []string
 	err = filepath.WalkDir(localPath, func(path string, d fs.DirEntry, err error) error {
 		if err == nil && d != nil {
-			log.Debug("FileCache::invalidateDirectory : %s (%d) getting removed from cache", path, d.IsDir())
+			// WalkDir goes through the tree in lexical order
+			// 	so 'aDir' always comes before 'aDir/file'
+			// 	and 'bDir' always comes after the whole 'aDir' subtree
+			// We can leverage this lexical order to queue folders for deletion after their children
+
+			// queue all files for deletion
 			if !d.IsDir() {
+				log.Debug("FileCache::invalidateDirectory : %s (file) getting removed from cache", path)
 				fc.policy.CachePurge(path)
-			} else {
-				deleteDirError := deleteFile(path)
-				if deleteDirError != nil {
-					log.Warn("FileCache::invalidateDirectory : %s - failed to delete %s. Here's why: %v", name, path, deleteDirError)
+			}
+			// queue folders for deletion after their children
+			if len(directoryStack) > 0 {
+				lastDirIndex := len(directoryStack) - 1
+				if !strings.HasPrefix(path, directoryStack[lastDirIndex]) {
+					log.Debug("FileCache::invalidateDirectory : %s (directory) getting removed from cache", path)
+					fc.policy.CachePurge(directoryStack[lastDirIndex])
+					directoryStack = directoryStack[:lastDirIndex]
 				}
 			}
+			// remember folders we've come across, to queue up later
+			if d.IsDir() {
+				directoryStack = append(directoryStack, path)
+			}
 		}
 		return nil
 	})
+	// queue up the remaining folders for deletion
+	for len(directoryStack) > 0 {
+		lastIndex := len(directoryStack) - 1
+		path := directoryStack[lastIndex]
+		log.Debug("FileCache::invalidateDirectory : %s (directory) getting removed from cache", path)
+		fc.policy.CachePurge(path)
```

Comment:
Will CachePurge handle directories for both lfu and lru? I thought that it only handled files, which is why delete was called separately for directories?

Reply:
For both policies, `CachePurge`  removes the path from the policy data structure, and then pushes the path into a queue for deletion.
Since folders are not considered to be cached items, when `CachePurge` goes to remove them from the policy records, it does nothing. But it still goes on to put them in the deletion queue.
The key to my bug fix here is that the folders need to be deleted in a queue - after their children - and not immediately (as they were doing before). So I am using `CachePurge` to achieve that without doing a lot of extra work creating a separate function.
Both policies implement `CachePurge` to ultimately call `os.Remove` on whatever is passed in. So I think this should work either way. But I have not tested it with LFU.

I honestly can't tell why they had different behavior for files and directories. Maybe it's related to the fact that part of the `CachePurge` functionality doesn't apply to directories... But it's honestly confusing, and I don't want to dissect the mistaken thinking behind it.

Reply:
Yeah I like the change, and the logic behind it. Maybe you could add a test for the LRU and LFU to test using CachePurge to delete a directory, to check that it works correctly.

Reply:
I have added tests for LRU and LFU.